src/backend/distributed/commands/multi_copy.c

```diff
@@ -1217,6 +1228,209 @@ ReportCopyError(MultiConnection *connection, PGresult *result)
+				getTypeInputInfo(destType, &iofunc, &(result->typioparam));
+				fmgr_info(iofunc, &(result->secondCoerceFunction));
+			}
+
+			return;
+		}
+
+		case COERCION_PATH_FUNC:
+		{
+			result->coercionType = COERCION_PATH_FUNC;
+			fmgr_info(coercionFuncId, &(result->firstCoerceFunction));
+			return;
+		}
+
+		case COERCION_PATH_RELABELTYPE:
+		{
+			result->coercionType = COERCION_PATH_RELABELTYPE;
+			return; /* the types are binary compatible, no need to call a function */
+		}
+
+		default:
+			Assert(false); /* there are no other options for this enum */
+	}
+}
+
+
+/*
+ * Returns the type of the provided column of the provided tuple. Throws an error if the
+ * column does not exist or is dropped.
+ *
+ * tupleDescriptor and relationId must refer to the same table.
+ */
+static Oid
+TypeForColumnName(Oid relationId, TupleDesc tupleDescriptor, char *columnName)
+{
+	AttrNumber destAttrNumber = get_attnum(relationId, columnName);
+	Form_pg_attribute attr = NULL;
+
+	if (destAttrNumber == InvalidAttrNumber)
+	{
+		ereport(ERROR, (errmsg("invalid attr? %s", columnName)));
+	}
+
+	attr = TupleDescAttr(tupleDescriptor, destAttrNumber - 1);
+	return attr->atttypid;
+}
+
+
+/*
+ * Walks a TupleDesc and returns an array of the types of each attribute. Will return
+ * InvalidOid in the place of dropped attributes.
+ */
+static Oid *
+TypeArrayFromTupleDescriptor(TupleDesc tupleDescriptor)
+{
+	int columnCount = tupleDescriptor->natts;
+	Oid *typeArray = palloc0(columnCount * sizeof(Oid));
+	int columnIndex = 0;
+
+	for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupleDescriptor, columnIndex);
+		if (attr->attisdropped)
+		{
+			typeArray[columnIndex] = InvalidOid;
+		}
+		else
+		{
+			typeArray[columnIndex] = attr->atttypid;
+		}
+	}
+
+	return typeArray;
+}
+
+
+/*
+ * ColumnCoercionPaths scans the input and output tuples looking for mismatched types,
+ * it then returns an array of coercion functions to use on the input tuples, and an
+ * array of types which descript the output tuple
+ */
+static CopyCoercionData *
+ColumnCoercionPaths(TupleDesc destTupleDescriptor, TupleDesc inputTupleDescriptor,
+					Oid destRelId, List *columnNameList,
+					Oid *finalColumnTypeArray)
+{
+	int columnIndex = 0;
+	int columnCount = inputTupleDescriptor->natts;
+	CopyCoercionData *coercePaths = palloc0(columnCount * sizeof(CopyCoercionData));
+	Oid *inputTupleTypes = TypeArrayFromTupleDescriptor(inputTupleDescriptor);
+	ListCell *currentColumnName = list_head(columnNameList);
+
+	for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
+	{
+		Oid destTupleType = InvalidOid;
+		Oid inputTupleType = inputTupleTypes[columnIndex];
+		char *columnName = lfirst(currentColumnName);
+
+		if (inputTupleType == InvalidOid)
+		{
+			/* this was a dropped column and will not be in the incoming tuples */
+			continue;
+		}
+
+		destTupleType = TypeForColumnName(destRelId, destTupleDescriptor, columnName);
```

Comment:
That's essentially what I'm doing, except instead of pulling the attribute out of the cache, I'm using the attribute which is already stored in destTupleDescriptor. I'll still need to call `get_attnum` though, since I only have a column name! I can swap the call to `TupleDescAttr` with a call to `get_atttype`, but not sure what's it wins us.